### ISSUE TITLE
Support for s3 bucket subdomains

### ISSUE THREAD

**Issue by aaronjwhiteside** [localstack/localstack#2631]
# Type of request: This is a ...
[x] feature request

# Detailed description
AWS has announced to deprecate pathStyleAccess in the foreseeable future.
https://aws.amazon.com/blogs/aws/amazon-s3-path-deprecation-plan-the-rest-of-the-story/

Consideration should also be given to libraries and software that cannot be configured with pathStyleAccess. We have such a use case where we cannot make this modification for lack of access to the original source code.

## Expected behavior
Edge service should be able to observe the incoming host header and use that to determine the bucket.

## Actual behavior
The edge service doesn't know what to do with the s3 request.

```
2020-06-27T03:40:36:INFO:localstack.services.edge: Unable to find forwarding rule for host "file-service-bucket.localstack:4566", path "/Dispute/2020/6/27/test.txt765966d6-64dd-4e7a-9db3-d82598f3fd65?uploads", target header "", auth header ""
```

This style of access can be supported by docker using network aliases, this is how I ran this example with the maven docker plugin..

```xml
                        <image>
                            <alias>localstack</alias>
                            <name>localstack/localstack:0.11.2</name>

                            <run>
                                <network>
                                    <name>${project.groupId}:${project.version}</name>
                                    <alias>localstack</alias>
                                    <alias>file-service-bucket.localstack</alias>
                                </network>

                                <env>
                                    <SERVICES>s3</SERVICES>
                                    <DEFAULT_REGION>us-west-1</DEFAULT_REGION>
                                    <DEBUG>1</DEBUG>
                                    <HOSTNAME>localstack</HOSTNAME>
                                    <EXTERNAL_HOSTNAME>localstack</EXTERNAL_HOSTNAME>
                                </env>

                                <ports>
                                    <port>${localstack.edge.port}:4566</port>
                                    <port>${localstack.s3.port}:4572</port>
                                </ports>

                                <volumes>
                                    <bind>
                                        <volume>${project.basedir}/src/test/docker-entrypoint-initaws.d:/docker-entrypoint-initaws.d</volume>
                                    </bind>
                                </volumes>

                                <wait>
                                    <log>make_bucket: file-service-bucket</log>
                                    <time>60000</time>
                                </wait>
                            </run>
                        </image>
```

┆Issue is synchronized with this [Jira Bug](https://localstack.atlassian.net/browse/LOC-205) by [Unito](https://www.unito.io/learn-more)

**Comment by jtomaszewski** [comment]
Once localstack can infer the s3 bucket from the host then the only missing part is to make it accessible under those hosts. If you know the bucket names in advance, you can do in your `docker-compose.yml` sth like this:

```
  localstack:
    image: localstack/localstack-light
    environment:
      - HOSTNAME_EXTERNAL=localstack
      - SERVICES=s3

  app:
    links:
      - localstack:ailo-files.localstack
```

Then both `localstack` and `ailo-files.localstack` hosts in `app` container point to `localstack`. 
Maybe that solution could be also then added to README as an example of how to make it work.

However, if the bucket name needs to be arbitrary (using wildcard), then I think it wont work then.

**Comment by JohanMollevik** [comment]
I am affected by this as well when trying to test unmodified docker containers on a private docker compose network.

Regarding wildcard names, I guess that if needed one could run a custom dns server in a docker internal network to route thous.

**Comment by jelhan** [comment]
> Once localstack can infer the s3 bucket from the host then the only missing part is to make it accessible under those hosts.

Is this something which should already work? I tried to set it up as described by you as I know the bucket name in advance but still running into connectivity issues:

```
cURL error 7: Failed to connect to my-bucket.localstack port 4556: Connection refused (see https://curl.haxx.
se/libcurl/c/libcurl-errors.html)
```

I have set `HOSTNAME_EXTERNAL` to the name of docker-compose service.

I also configured the `links` as described by you. I'm sure that this is working. Before I was seeing this error: `Could not resolve host: my-bucket.localstack`. It was resolved by adding the alias as described by you.

**Comment by luqasn** [comment]
I was trying to do the exact same thing as @JohanMollevik and came up with a combination of mitmproxy, CoreDNS and Traefik to allow testing unmodified images against AWS.
Please see the (very fresh) project at https://github.com/luqasn/aws-sandbox and tell me what you think.

It provides both AWS dualstack compatibility as well as domain style access via custom DNS and mitmproxy rewriting.

**Comment by karlismelderis** [comment]
> I was trying to do the exact same thing as @JohanMollevik and came up with a combination of mitmproxy, CoreDNS and Traefik to allow testing unmodified images against AWS.
> Please see the (very fresh) project at https://github.com/luqasn/aws-sandbox and tell me what you think.
> 
> It provides S3 domain style access with dualstack name format support via custom DNS and mitmproxy rewriting.

👍 thank you for inspiration

**Comment by yh1224** [comment]
The deprecation of Path-style seems to have been postponed for at least one year.
https://aws.amazon.com/blogs/aws/amazon-s3-path-deprecation-plan-the-rest-of-the-story/

**Comment by rilaby** [comment]
Basically, the plan for path-style navigation has changed.
**tl;dr:** all buckets created after September 2020 can't use path-style whereas all the others (created before that date) could use path-style forever (or until another notice from Amazon).

Saying that I would like to raise this issue up.

**Comment by macnev2013** [comment]
Thanks for reporting. This feature was added in PR #3690. Please pull the `latest` image of the docker using `docker pull localstack/localstack:latest` and give it a try?
Thanks.

**Comment by sats17** [comment]
Hi @macnev2013 ,

Do we require any configuration changes to use this feature ?

Thanks.

**Comment by BeyondMySouls** [comment]
Hi,

Indeed, been struggling with this for some days since this feature was released but it did not work for me either.

**Comment by whummer** [comment]
Thanks for the update @sats17 @BeyondMySouls . Generally, the new feature (S3 bucket subdomains, i.e., virtual host based bucket addressing) should be working without configuration changes, as the code is still backward compatible with path-based addressing.

> Error - Could not resolve host: bucketname.localhost
Can you please share some details about your environment setup - where exactly are you seeing this error, and which action / API call triggers the error? Thanks!

**Comment by sats17** [comment]
Hi @whummer,

My docker-compose.yml have configurations which in below gist link.
https://gist.github.com/sats17/1e45e717d68ef271f43d394625de4bcd

I was trying to create S3 snapshot repo from elastic search (6.7.0). I have done all prerequisites like bucket creation, installing s3 plugin on elasticsearch and setting up access key and secret key.
Following curl command which I am using to elasticsearch.
```
curl -X PUT "http://localhost:9200/_snapshot/test_repo" -H 'Content-Type: application/json' -d'
{
     "type": "s3",
     "settings": {
         "bucket": "mybucket",
         "endpoint": "http://localhost:4566/"
    }
}'
```
And I am getting following exception from elasticsearch

```
{
"error": {
"root_cause": [
{
"type": "repository_verification_exception",
"reason": "[mybucket] path is not accessible on master node"
}
],
"type": "repository_verification_exception",
"reason": "[mybucket] path is not accessible on master node",
"caused_by": {
"type": "i_o_exception",
"reason": "Unable to upload object [tests-aMNU5vkYTa6Mb8LX-jN2wQ/master.dat] using a single upload",
"caused_by": {
"type": "sdk_client_exception",
"reason": "Unable to execute HTTP request: mybucket.localhost",
"caused_by": {
"type": "unknown_host_exception",
"reason": "mybucket.localhost"
}
}
}
},
"status": 500
}
```
Also I tried to do simple curl request on localstack but it didn't worked.
```
curl -X GET "http://mybucket.localhost:4566"
Could not resolve host: mybucket.localhost
```

Thanks..

**Comment by BeyondMySouls** [comment]
On my side, `curl -X GET "http://my-bucket.localhost:4566"` failed but it did work with the other endpoint: `curl -X GET "http://my-bucket.s3.localhost.localstack.cloud:4566"`

**Comment by sats17** [comment]
Hi @whummer ,

Yes it is working with endpoint http://s3.localhost.localstack.cloud:4566

**Comment by yogevyuval** [comment]
@sats17 @whummer @BeyondMySouls @macnev2013 
Is there a solution when this set is running inside docker containers? Meaning localstack runs as a container, but also the app that is accessing it is running in a container. In this situation we have to use the container name in order to communicate with localstack, which doesnt work with the combination of virtual host paths.

**Comment by kevinscholz** [comment]
> Is there a solution when this is running inside docker containers? Meaning localstack runs as a container, but also the app that is accessing it is running in a container. In this situation we have to use the container name in order to communicate with localstack, which doesnt work with the combination of virtual host paths.

You might want to have a look at network aliases. I use this config to run localstack in Docker Compose, e.g. for my bucket `test-bucket`:

```
  s3:
    image: localstack/localstack:0.12.9
    hostname: s3
    networks:
      default:
        aliases:
          - s3.localhost.localstack.cloud
          - s3-website.localhost.localstack.cloud
          - test-bucket.s3.localhost.localstack.cloud
          - test-bucket.s3-website.localhost.localstack.cloud
    environment:
      SERVICES: s3
      DEFAULT_REGION: eu-central-1
      DEBUG: "1"
    expose:
      - "4566"

```

**Comment by harveyappleton** [comment]
@yogevyuval I'm running a NodeJS application in one docker container, and localstack in another container.
I've bridged both to the host (so can access either via `localhost:5000` for node or `localhost:4566` for localstack).

I've set in my s3 client config:
```
{ forcePathStyle: true, endpoint: http://host.docker.internal:4566, ... }
```

so that subdomains are not an issue for my docker setup.
(However, wondering if the AWS deprecation will affect `forcePathStyle` functionality? No deprecation notices inside of TS definitions though...)

**Comment by mattsoftware** [comment]
So for the life of me I cannot seem to get this working. If I run localstack in a docker container, and access it using <bucket>.s3.localhost.localstack.cloud:4566 I can get it working, but I run a couple of localstack containers for different projects that could all be running at the same time.

For example I might have a localstack, available on 127.0.0.4:4566 and have updated my hosts file to say something like this...
127.0.0.4 project.aws.local

When I want to connect and make localstack aws calls, i set the endpoint to http://project.aws.local:4566 and everything works great.

Now, how can I access using s3 paths? I have tried adding s3.project.aws.local and <bucketname>.s3.project.aws.local to my hosts file, but none of them seem to work. It leads me to believe it only works under the localhost.localstack.cloud endpoint. I have tried updating the LOCALSTACK_HOSTNAME and HOSTNAME_EXTERNAL in my docker-compose.yml file, as well as setting network aliases on the container, as well as updating my hosts file so I can hopefully access from my localhost as well, but I really cant seem to find a config that will work for me. 

Any idea's or a step-by-step to work out how to get this working?

**Comment by arthaszeng** [comment]
> > Is there a solution when this is running inside docker containers? Meaning localstack runs as a container, but also the app that is accessing it is running in a container. In this situation we have to use the container name in order to communicate with localstack, which doesnt work with the combination of virtual host paths.
> 
> You might want to have a look at network aliases. I use this config to run localstack in Docker Compose, e.g. for my bucket `test-bucket`:
> 
> ```
>   s3:
>     image: localstack/localstack:0.12.9
>     hostname: s3
>     networks:
>       default:
>         aliases:
>           - s3.localhost.localstack.cloud
>           - s3-website.localhost.localstack.cloud
>           - test-bucket.s3.localhost.localstack.cloud
>           - test-bucket.s3-website.localhost.localstack.cloud
>     environment:
>       SERVICES: s3
>       DEFAULT_REGION: eu-central-1
>       DEBUG: "1"
>     expose:
>       - "4566"
> ```

THANKS! life saver

**Comment by evbo** [comment]
Having to hard-code bucket names into docker-compose network aliases [as shown](https://github.com/localstack/localstack/issues/2631#issuecomment-826867596) by @kevinscholz doesn't seem like a scalable answer when many s3 buckets are involved and defined in terraform; One must refactor docker-compose each time the terraform names change for the s3 buckets.

So can this issue be reopened to provide network access to s3 buckets over the `localstack` docker compose network that gets generated when running localstack as a docker compose service?

**Comment by bentsku** [comment]
Hi @evbo,

This is a limitation of Docker network, and they don't support wildcard addressing for network aliases at the moment. One way to get around this limitation is to use path style when using LocalStack, as shown here: https://docs.localstack.cloud/user-guide/integrations/terraform/#request-management

Do you have a hard requirement on virtual host style?

**Comment by evbo** [comment]
Thanks @bentsku but, due to AWS deprecating path style, won't that no longer be an option at some point?

I don't have any great suggestions either, but one idea:

- Add a minimal and optional DNS proxy client library that will [read wildcard entries in etc/hosts](https://serverfault.com/a/265396) 
- Then, users install it in their client docker containers: `apt-get install localstack-dns-client`
- maybe docker compose's `extra_hosts` configuration can be used to add wildcard entries? If not users would need to directly add them to /etc/hosts file. Or maybe the proxy client could do this?

**Comment by bentsku** [comment]
AWS is deprecating it, but it will still be supported for a while, as there's no support for virtual host bucket with dots in the name for now, due to SSL certificates. As long as the SDKs allow to be configured for path addressing, we can continue to use this.

But to have some networks utils to be able to use virtual host addressed buckets between containers would for sure be nice, I completely agree. Pinging @simonrw in this discussion as he's currently working on LocalStack networking, on a different subject for now but related.

**Comment by SleepyBrett** [comment]
Hey guys, I ran into this in a vcluster with a localstack and for anyone that came after me I just went ahead and edited the coredns config file and added this line:

```
rewrite name regex .*\.localstack\.default\.svc\.cluster\.local localstack.default.svc.cluster.local
```

make the changes you need to make regarding the namespace and service name and whatnot.

**Comment by arthaszeng** [comment]
您好，你发的邮件已收到，我会尽快给您回复，谢谢。
